### PR TITLE
elf: replace .got.zig with a zig jump table

### LINKED ISSUE #20887: rethink `.got.zig`
follow-up on #20884

After #20884, we are no longer compatible with existing tooling as custom Zig relocs are emitted for RISC-V when using `build-obj`. In this specific case, we needed to create a relocation for statically accessing the GOT in a non-pcrel manner. However, there are still flaws in the `.got.zig` design, like the self-hosted backends need to create logic to handle whether a symbol is inside of a `.got.zig` or not.

Microsoft thought up a neat idea we could apply instead called a thunk table. It's used in `link.exe` and is probably the only functional incremental linking implementation. @kubkon and I would like to replace `.got.zig` with it one day since it plays nicer with other arches.

### PR DESCRIPTION
**Pull request by kubkon**
Motivating factor: make this feature as transparent to the codegen as possible

Closes #20887 

Previously, we would use `.got.zig` to indirect pointers to global data too but as was agreed on many an occasion we only really want to indirect function calls via an offset table or similar. In fact, as far as I understand that was the original plan of @andrewrk when he wrote the first PoC of incremental ELF linker. Therefore, since we longer want to indirect pointers to global data, it makes sense to replace an offset table with an equivalent jump (trampoline) table that is directly embedded within the machine code section. This improves code locality but also should make load times shorter if dynamically linking since we no longer have to rebase any pointers.

The new jump table looks as follows (for x86_64):
```
<0>: jmp symbol_a // jmp rel32
<5>: jmp symbol_b
...
```

Compared to storing pointers, if we have to relocate the table because it outgrew its capacity, we will have to re-calculate the jump targets since the jump sequence is PC-relative, however, this can be reduced into applying a fixed offset to every entry when relocating.

From the perspective of the codegen, when emitting a call with a relocation the codegen no longers needs to care about the presence or absence of the jump table/offset table - it simply emits `call rel32` with `R_X86_64_PLT32` relocation where the target is `symbol_a`. Then, the linker upon resolving `R_X86_64_PLT32` will check if the jump table has been created and the symbol can be indirected via said table and rewrite the target address to the jump table entry if so. Again, this is all transparent to the codegen. As an added bonus, codegen now generates identical code in `build-exe` and `build-obj` modes.

One caveat of the new approach is that we only indirect function calls - if you request a function pointer, currently you will receive exactly that, with no indirection.

I am looking forward to the feedback if we should proceed or whatnot!

TODO
- [x] x86_64 codegen
- [x] x86_64-elf trampolines
- [x] riscv codegen
~~- [ ] riscv-elf trampolines~~ (deferred until we have a working incremental linker)

### CODE REVIEW

**Comment by andrewrk** [comment]
> One caveat of the new approach is that we only indirect function calls - if you request a function pointer, currently you will receive exactly that, with no indirection.

Couldn't the function pointer point to the address in the jump table? Then the pointer would be also correct after that jump table address has been updated.

**Comment by jacobly0** [comment]
> > One caveat of the new approach is that we only indirect function calls - if you request a function pointer, currently you will receive exactly that, with no indirection.
> 
> Couldn't the function pointer point to the address in the jump table? Then the pointer would be also correct after that jump table address has been updated.

That would not support how `@alignCast` is currently implemented on function pointers, such as here https://github.com/ziglang/zig/blob/0b5ea2b902b5802786cac70740e93872d2a0973d/test/behavior/align.zig#L369

**Comment by andrewrk** [comment]
I don't see the problem.

**Comment by mlugg** [comment]
@jacobly0, I don't understand how `@alignCast` would be affected here -- the example you link to is just taking an `align(4)` function, casting the pointer down to `align(1)`, then casting it back up to `align(4)`. Of course, if we made the function pointer point into the jump table, then the jump table entries would themselves need to be aligned according to the function's alignment, but that seems perfectly doable to me. Am I missing something?

**Comment by jacobly0** [comment]
The way alignment works, in order to align a part of a whole, the whole also has to be at least that aligned, which means you would need the max alignment up front for the jump table itself, but the whole point is that these entries don't move and that new, higher alignments could come in later.

**Comment by andrewrk** [comment]
The jump table entries only need to be aligned to the address size.

**Comment by andrewrk** [comment]
Ah, I see the problem, it's that Zig allows setting function alignment. That can be dropped from the language. It's not useful and as you can see it restricts the compiler from lowering in certain useful ways.

**Comment by kubkon** [comment]
> Ah, I see the problem, it's that Zig allows setting function alignment with `align(X)` syntax. That can be dropped from the language. It's a stretch to call it useful, it's not portable, and as you can see it restricts the compiler from lowering in certain useful ways.

After chatting to Jacob about it, he suggested we re-use the current address of the function as its trampoline (think distributed jump table of sorts) - this brilliant idea of his effectively solves the function alignment problem with one rare edge case, namely if we further overalign the function. However, in this case since it's rather rare it will most likely be fine to just update all referring symbols thus skipping creating a trampoline.

I have implemented this idea as part of this PR and it looks like this:

```zig
if (old_rva != new_rva and old_rva > 0) {
        // If we had to reallocate the function, we re-use the existing slot for a trampoline.
        // In the rare case that the function has been further overaligned we skip creating a
        // trampoline and update all symbols referring this function.
        if (old_alignment.order(new_alignment) == .lt) {
            @panic("TODO update all symbols referring this function");
        }

        // Create a trampoline to the new location at `old_rva`.
        if (!self.symbol(sym_index).flags.has_trampoline) {
            const name = try std.fmt.allocPrint(gpa, "{s}$trampoline", .{
                self.symbol(sym_index).name(elf_file),
            });
            defer gpa.free(name);
            const name_off = try self.addString(gpa, name);
            const tr_size = trampolineSize(elf_file.getTarget().cpu.arch);
            const tr_sym_index = try self.newSymbolWithAtom(gpa, name_off);
            const tr_sym = self.symbol(tr_sym_index);
            const tr_esym = &self.symtab.items(.elf_sym)[tr_sym.esym_index];
            tr_esym.st_info |= elf.STT_OBJECT;
            tr_esym.st_size = tr_size;
            const tr_atom_ptr = tr_sym.atom(elf_file).?;
            tr_atom_ptr.value = old_rva;
            tr_atom_ptr.alive = true;
            tr_atom_ptr.alignment = old_alignment;
            tr_atom_ptr.output_section_index = elf_file.zig_text_section_index.?;
            tr_atom_ptr.size = tr_size;
            const target_sym = self.symbol(sym_index);
            target_sym.addExtra(.{ .trampoline = tr_sym_index }, elf_file);
            target_sym.flags.has_trampoline = true;
        }
        const target_sym = self.symbol(sym_index);
        try writeTrampoline(self.symbol(target_sym.extra(elf_file).trampoline).*, target_sym.*, elf_file);
    }
}
```

**Comment by kubkon** [comment]
Oh and forgot to mention. A rather fortuitous byproduct of Jacob's idea is that for a one-off (or non-incremental) case, no trampoline will be created thus reducing the binary size.

**Comment by mlugg** [comment]
Just to check I'm not missing something: that idea doesn't quite work with binary hot-swapping, right? Since the current instruction pointer may be at the start of a function, so rewriting the first N bytes to a trampoline would potentially result in reading bogus instructions. Although, I guess this could easily be fixed by having the first instruction of every function be a "trampoline" (jmp a few bytes ahead!) to the directly-succeeding function code in hot-swappable builds; so your first instruction is always a jmp, it just might be to the old function if pre-hotswap or to the new function if post-hotswap.

**Comment by kubkon** [comment]
> Just to check I'm not missing something: that idea doesn't quite work with binary hot-swapping, right? Since the current instruction pointer may be at the start of a function, so rewriting the first N bytes to a trampoline would potentially result in reading bogus instructions. Although, I guess this could easily be fixed by having the first instruction of every function be a "trampoline" (jmp a few bytes ahead!) to the directly-succeeding function code in hot-swappable builds; so your first instruction is always a jmp, it just might be to the old function if pre-hotswap or to the new function if post-hotswap.

That's a fair point. I wonder if `nop`ing the trampoline size at the beginning of each function would also help. Not sure, but when we get there we have enough options to explore and make it work I believe.

**Comment by kubkon** [comment]
Another issue to have in mind is that the function itself cannot be smaller than the trampoline size (with padding ofc), but this is an easy fix also.

**Comment by kubkon** [comment]
> > That's a fair point. I wonder if `nop`ing the trampoline size at the beginning of each function would also help. Not sure, but when we get there we have enough options to explore and make it work I believe.
> 
> The problem there is that if the arch doesn't have a single wide NOP of the jump instruction size, the instruction pointer might be in one of the non-initial NOPs, so again we read invalid instructions. That's why I suggested leading with a JMP to a few bytes ahead -- it doesn't even matter if this instruction is smaller (e.g. a short relative jump on x86_64) since the IP will always jump from it straight to the start of the "real" code.

Is there an arch like this? Also, this still might be a problem for RISC where a trampoline would comprise multiple instructions such as

```
adrp label
add label
br
```

wouldn't it?

**Comment by andrewrk** [comment]
I still think a jump table is better:
* Avoid garbage in CPU cache that would otherwise occur when a function is moved leaving behind its trampoline followed by unused space
* Avoid fragmentation in virtual memory space and the object file that would otherwise occur when a function is moved leaving behind its trampoline interrupting unused space
* There is no function pointer alignment problem. Everything is fine.
* Easier bookkeeping; simpler data structure for tracking the trampolines.

**Comment by kubkon** [comment]
> I still think a jump table is better:
> 
>     * Avoid garbage in CPU cache that would otherwise occur when a function is moved leaving behind its trampoline followed by unused space
> 
>     * Avoid fragmentation in virtual memory space and the object file that would otherwise occur when a function is moved leaving behind its trampoline interrupting unused space
> 
>     * There is no function pointer alignment problem. Everything is fine.
> 
>     * Easier bookkeeping; simpler data structure for tracking the trampolines.
> 
>     * Simpler hot code swapping, simpler incremental linking

How is function pointer alignment problem solved with a jump table?

**Comment by andrewrk** [comment]
Function pointers are always pointer-size aligned when using a jump table.

**Comment by kubkon** [comment]
> Function pointers are always pointer-size aligned when using a jump table.

I guess I should have asked this first: do we indirect pointers to functions via a jump table too? If so, this will not currently succeed:

```zig
test "align(N) on functions" {
    try expect((@intFromPtr(&overaligned_fn) & (0x1000 - 1)) == 0);
}

fn overaligned_fn() align(0x1000) i32 {
    return 42;
}
```

**Comment by kubkon** [comment]
Also please note that every entry in a jump table is not pointer-size aligned, but instruction aligned. Perhaps you are referring to an offset table?

**Comment by andrewrk** [comment]
> garbage in CPU cache

This is easy to understand:

Here is a cache line

```
[ ] [ ] [ ] [ ] [ ] [ ] [ ] [ ] 
```

Here is a cache line full of jump table data:

```
[F] [F] [F] [F] [F] [F] [F] [F] 
```

All those F's are valid pointers to functions that might be used.

Here is a cache line after a function has been relocated with the other strategy:

```
[T] [T] [J] [T] [T] [T] [T] [T] 
```

J - jump instruction to the real function
T - trash. 100% guaranteed waste of space in the CPU cache

**Comment by kubkon** [comment]
> > garbage in CPU cache
> 
> This is easy to understand:
> 
> Here is a cache line
> 
> ```
> [ ] [ ] [ ] [ ] [ ] [ ] [ ] [ ] 
> ```
> 
> Here is a cache line full of jump table data:
> 
> ```
> [F] [F] [F] [F] [F] [F] [F] [F] 
> ```
> 
> All those F's are valid pointers to functions that might be used.
> 
> Here is a cache line after a function has been relocated with the other strategy:
> 
> ```
> [T] [T] [J] [T] [T] [T] [T] [T] 
> ```
> 
> J - jump instruction to the real function T - trash. 100% guaranteed waste of space in the CPU cache

Ok so it seems you mean an offset table being a better solution than a jump table be it in one big block or distributed.

**Comment by andrewrk** [comment]
> Actually bookkeeping is largely unchanged whether the jump table is as one big block or distributed. If a function has a trampoline, it gets an `extra` slot with an index to the symbol that acts as its trampoline. In case of a jump table, it gets an index into a jump table.

This means that with the function header trampoline strategy, you have to keep around an old symbol data entry around for the now-deleted function, whereas with the jump table strategy, you don't.

**Comment by andrewrk** [comment]
> Ok so it seems you mean an offset table being a better solution than a jump table be it in one big block or distributed.

No, my mistake for saying "function pointer" when I should have said "jump instruction" but otherwise the same point stands.

It looks like on x86_64, jumps with absolute address is 5 bytes which is annoying, but it's fine. Function pointers will have alignment of 1 on that platform then.

This test case will regress:

```zig
test "align(N) on functions" {
    try expect((@intFromPtr(&overaligned_fn) & (0x1000 - 1)) == 0);
}

fn overaligned_fn() align(0x1000) i32 {
    return 42;
}
```

The language will be modified to say when you take the address of a function, it does not necessarily gain the machine code alignment specified with the `align` keyword, because there may be a stub being used, or something to this effect.

Even if we go with the function prologue strategy resolution, I will still make this language change, because it is already evident this flexibility is useful for compilers.

**Comment by kubkon** [comment]
> > Actually bookkeeping is largely unchanged whether the jump table is as one big block or distributed. If a function has a trampoline, it gets an `extra` slot with an index to the symbol that acts as its trampoline. In case of a jump table, it gets an index into a jump table.
> 
> This means that with the function header trampoline strategy, you have to keep around an old symbol data entry around for the now-deleted function, whereas with the jump table strategy, you don't.

I am getting confused, let's agree on what is what if that's OK. We have 3 options under consideration:
1. Offset table
```
read-write section, each entry pointer-aligned, pointer-size
<0>: pointer to A
<8>: pointer to B
...
```
2. Jump table
read-exec section, `.text.zig`, each entry is instruction-aligned (1 on x86_64), size is arch dependent, code model dependent, etc
<0>: jmp rel32 // 5 bytes, PC-relative near-jump
<5>: ...
3. Distributed jump table
read-exec section, `.text.zig`, each entry is at least instruction-aligned or overaligned if the evacuated function was moved
<0>: code of A
...
<N>: jmp rel32 // jump to now-moved B
<N+5>: free space
....
<M>: code of C etc
```

FWIW I think there is a way to keep option 1, or at least reconsider it, while ensuring codegen backends are largely incremental linking agnostic by utilising the idea of lazy-symbol binding - we emit a section with immutable jump entries that always point to the same pointer in the offset table. Offset table is again mutable where we are free to rewrite pointers, however since the machine code now points at an immutable jump table we keep the codegen largely unaware of incremental linking. It would look something like this:

```
// .text.zig
<0>: A
...
<N>: call B -> .plt.zig, .stub.zig, or whatever
...

// .plt.zig
<0>: mov r11, [.got.zig entry at #0]
<N>: call [r11]
...

// .got.zig
<0>: pointer to B
<8>: pointer to A
...
```

**Comment by kubkon** [comment]
FWIW the obvious con of option 4* is having to (re-)introduce two more sections `.got.zig` and `.plt.zig` which is costly on MachO where we are severely limited in section numbers.

**Comment by andrewrk** [comment]
I think we should toss out option 1 purely on a performance basis:
* Offset table: every function call is indirect (20-50 cpu cycles)
* Jump table (distributed or not): every function call is directly (15-30 cpu cycles)

[source](http://ithare.com/infographics-operation-costs-in-cpu-clock-cycles/)

The discussion right now is jump table vs distributed jump table. I don't think anyone is advocating for offset table.

**Comment by kubkon** [comment]
Oh I brought up option 4 to point out yet another mechanism at our disposal in case it wasn't clear how an offset table can be utilised while making codegen agnostic to the concept of indirection.

**Comment by jacobly0** [comment]
If you pack all of the jumps together, you enforce that every new function call requires a minimum of 2 cache references, 3 if the jump instruction itself crosses a cache boundary (since 5 is not a power of 2, and pretty terrible since it would happen on a whopping 6% of functions).  In order to fully utilize the first cache reference, you have to get lucky with using all the adjacent jump entries at around the same time, which is not even slightly reasonable to expect.

If you prefix the majority of non-changing functions with their own jump, then you allow them to only require a single cache reference that is fully utilized for all but the most trivial of functions.  Saying that this space goes to waste after moving the function is misleading because it could be filled with any other function that fits in that space, in the same way that unused jump entries can be filled with any other jump.  Even in the split case, you only have to get lucky with two functions being in use at the same time to fully utilize both caches.

Increasing the minimum cost of all functions is clearly inferior to only increasing the cost of _some_ changing functions (those rapidly increasing in size).  This is even more true when you realize that once a split function stops changing over a period of time, it can be reunited in a "garbage collection"-like manner, restoring the original performance after you are happy with the new function implementation and stop editing it (in a way that vastly increases size).  This trades the cost of updating all references every time a function increases in size, to only doing it once the function has stabilized and stops increasing in size.  This is never possible with a jump table because those jumps can never be adjacent to their implementation.

The bookkeeping seems completely equivalent, for each symbol you either track where the jump table entry is or which possibly not yet named symbol contains the actual implementation (something which can also be trivially recovered by just reading the jump instruction).

I'm also not sure why we are intent on removing function pointer alignment from the language, since once a function is aligned to a cache line, there is little benefit to aligning it any more other than gaining bits in the function pointer to be used for other things.  It's the alignment of the first jump that matters in the common non-split jump prologue case, aligning the "beginning" of the function would just move it to a different cache line negating any benefits.

**Comment by andrewrk** [comment]
Thanks for chiming in @jacobly0. I'm convinced by your performance-related arguments.

> I'm also not sure why we are intent on removing function pointer alignment from the language, since once a function is aligned to a cache line, there is little benefit to aligning it any more other than gaining bits in the function pointer to be used for other things. It's the alignment of the first jump that matters in the common non-split jump prologue case, aligning the "beginning" of the function would just move it to a different cache line negating any benefits.

I don't understand what you're saying here. It seems like you're making an argument against function pointer alignment being useful, which seems to comport with removing the alignment guarantees of function pointers. But your conclusion is that we should keep function pointer alignment guarantees?

**Comment by jacobly0** [comment]
I'm saying that function alignment doesn't have much use at all without function pointer alignment, and so I don't understand the stance that only one should be removed.  I'd also argue that its usage is niche enough that theorizing about stubs is not very relevant given that stubs can also be aligned and almost no functions have an explicit alignment.  I think Zig already made the correct choice by making functions pointers default to `align(1)` instead of the target-specific default alignment that is actually used on functions to allow this flexibility in implementation by default.

**Comment by andrewrk** [comment]
I see, so, would you be for or against completely removing align(N) syntax from function declarations then?

**Comment by jacobly0** [comment]
I don't have a strong opinion on the outcome, I just have a strong opinion against using the discussion in this thread to justify removal.  If other arguments are made for its removal, I could probably be easily convinced, but as of today, it still seems like an experiment worth keeping.

**Comment by andrewrk** [comment]
Understood, thank you for the clarifications. I'm satisfied with the distributed jump table solution then.

File: src/arch/x86_64/CodeGen.zig [hunk 1]
```diff
@@ -1379,14 +1379,19 @@ fn asmImmediate(self: *Self, tag: Mir.Inst.FixedTag, imm: Immediate) !void {
         .ops = switch (imm) {
             .signed => .i_s,
             .unsigned => .i_u,
+            .reloc => .rel,
+        },
+        .data = switch (imm) {
+            .reloc => |x| .{ .reloc = x },
```
Comment (jacobly0) [inline src/arch/x86_64/CodeGen.zig:1385]:
```suggestion
            .reloc => |x| reloc: {
                assert(tag[0] == ._);
                break :reloc .{ .reloc = x };
            },
```

**Review by jacobly0 — APPROVED** [review]

File: src/link/Elf/Atom.zig [hunk 2]
```diff
@@ -2230,7 +2171,7 @@ const riscv = struct {
     const riscv_util = @import("../riscv.zig");
 };
 
-const ResolveArgs = struct { i64, i64, i64, i64, i64, i64, i64, i64 };
+const ResolveArgs = struct { i64, i64, i64, i64, i64, i64, i64 };
```
Comment (andrewrk) [inline src/link/Elf/Atom.zig:2174]: why is this a tuple instead of a struct?

**Review by andrewrk — APPROVED** [review]

File: src/link/Elf/Atom.zig (hunk 2 above)
Comment (kubkon) [reply src/link/Elf/Atom.zig:2174]: It was like this before but I ought to make it a struct instead of a tuple. Will do that in a follow-up.

**Pull request merged**